SharpBucket/V2/EndPoints/BranchResource.cs

```diff
@@ -25,10 +25,28 @@ public BranchResource(string accountName, string repository, RepositoriesEndPoin
         /// <summary>
         /// Lists all branches associated with a specific repository.
         /// </summary>
+        /// <param name="max">The maximum number of items to return. 0 returns all items.</param>
         /// <returns></returns>
-        public List<Branch> ListBranches()
```

Comment:
Removing that signature is a runtime breaking change. That means that a rebuilt is required
I understand why you did it: try to have the same style of signatures on all routes.
But since it seems hard to achieve without creating a runtime breaking change in any class, may be we should take a more radical direction and provide a single method with a similar signature on all classes where filter, sort and max parameters are all optional ?

We will just have the following signatures to dispatch in the appropriate classes:
`public List<Branch> ListBranches(string filter = null, string sort = null, int max = 0)`
`public List<PullRequest> ListPullRequests(string filter, string sort, int max = 0)`
`public List<Repository> ListRepositories(string accountName, string filter, string sort, int max = 0)`
`public List<Tag> ListTags(string filter, string sort, int max = 0)`

It's still a runtime breaking changes, but less code to maintain, and the ability to specify the sort without explicitly provide the filter

And since there is no reasons for SharpBucket to limit ListPullRequests to 100 by default and not 0 like other list methods, since we already introduce a runtime breaking change, it seems reasonable for me to fix that at the same time (even if this is more a compilation breaking change since code should be modify by people who expect this to be 100)

Reply:
I did change `ListBranches()` to `ListBranches(int max = 0)`, to stay consistent with the other resources, but I didn't thinkit was a breaking change, since any user calls to `ListBranches()` will still work.

Regarding your suggestion to unify the signatures, in general I don't like methods with lots of optional parameters; I prefer multiple overloads that get you closer to what you actually want. Otherwise you can wind up with a call like `ListBranches(null, null, 5)` just to get the max value in there. (I kept the `max` as an optional parameter in the overloads I added just to stay consistent with what was already there.)

If we now have three optional parameters, it might be better to move them into a single `ListParameters` object, and then have two overloads for each method, like so:

```csharp
class ListParameters
{
    public string Filter { get; set; }
    public string Sort { get; set; }
    public int Max { get; set; }
}

public List<Branch> ListBranches();
public List<Branch> ListBranches(ListParameters parameters);
```
What do you think?

Reply:
Since C# 4 we can call methods with named parameters, which mean that `ListBranches(null, null, 5)` can be write `ListBranches(max: 5)` which make calls to methods with lot of optional arguments easy and really expressive.

But the next time we will want to add another argument we will be against in front of that breaking change issue...

Your suggestion of creating a ListParameters class is really interesting because we will be able to add more optional parameters to that set of methods without generating any type of breaking change

So go for your last suggestion with a ListParameter class, I think it's the best one we have.

Reply:
Since you suggest to create two signatures:
```CSharp
public List<Branch> ListBranches();
public List<Branch> ListBranches(ListParameters parameters);
```

I would be favorable that the second one will check that the provided ListParameters instance is not null.
Which means that `ListBranches()` is not equivalent to `ListBranches(null)`, but equivalent to `ListBranches(new ListParameters())`

I think it can avoid issues where people declare a parameters variable but forget to assign it.

Is it what you have in mind to ?
What do you think of that ?

Reply:
Do you mean a null check that will throw an `ArgumentNullException`?

Reply:
Yes it's what I've in mind. What do you think of that ?

Reply:
My original thought was not to do a null check, but I can see that it makes sense if we're keeping both method signatures. I'll do it that way.